src/lib/core/utils/study-metadata.ts

```diff
@@ -31,3 +31,25 @@ export function makeEntityDisplayName(entity: StudyEntity, isPlural: boolean) {
     ? entity.displayName
     : entity.displayNamePlural ?? `${entity.displayName}s`;
 }
+
+export function findCollections(
+  entity: StudyEntity,
+  collections: Array<any> = []
+) {
+  if (entity.collections?.length) {
+    collections.push(
+      entity.collections.map((collection) => {
+        collection.entityId = entity.id;
+        collection.entityDisplayName = entity.displayName;
+        return { entityId: entity.id, variableId: collection.id };
+      })
+    );
+  }
+  if (entity.children?.length) {
+    entity.children.forEach((childEntity) =>
+      findCollections(childEntity, collections)
+    );
+  }
+
+  return collections;
+}
```

Comment:
The overall logic looks sound - looks like you're doing a pre-order traversal of the study entities and "transforming" each  `collection` into a form the plugin needs.

Instead of "rolling your own" preorder traversal of the study entities, I would use our existing `preorder` utility from WDKClient to collect all the study entities into an array. (Example use [here](https://github.com/VEuPathDB/web-eda/blob/413b0ef2bcbe7e5fd15277670cb875570564c1f9/src/lib/mapveu/MapVeuAnalysis.tsx#L26-L28).) You could then use array methods like `map` / `flatMap` to transform each `collection` into whatever form the plugins expect.